File: src/core/differ.ts
import { LRU } from '../utils/lru.js'
import { hashNormalized, hashObject } from './hasher.js'

/**
 * Tracks content hashes across requests to figure out what's stable
 * (same every call) vs what changes (user query, conversation tail).
 *
 * The idea: if a content block appears in 2+ of your last N requests,
 * it's almost certainly stable and worth caching.
 */
export class ContentDiffer {
  private seen: LRU<string, number>
  private threshold: number

  constructor(windowSize: number = 100, stabilityThreshold: number = 2) {
    this.seen = new LRU(windowSize)
    this.threshold = stabilityThreshold
  }

  /**
   * Record a content block and return how many times we've seen it.
   */
  recordAndCount(content: string): number {
    const hash = hashNormalized(content)
    const count = (this.seen.get(hash) || 0) + 1
    this.seen.set(hash, count)
    return count
  }

  recordObjectAndCount(obj: unknown): number {
    const hash = hashObject(obj)
    const count = (this.seen.get(hash) || 0) + 1
    this.seen.set(hash, count)
    return count
  }

  /**
   * Check if we've seen this content enough times to consider it stable.
   */
  isStable(content: string): boolean {
    const hash = hashNormalized(content)
    return (this.seen.get(hash) || 0) >= this.threshold
  }

  isObjectStable(obj: unknown): boolean {
    const hash = hashObject(obj)
    return (this.seen.get(hash) || 0) >= this.threshold
  }

  clear(): void {
    this.seen.clear()
  }
}

File: python/cachellm/core/differ.py
from cachellm.utils.lru import LRU
from cachellm.core.hasher import hash_normalized, hash_object


class ContentDiffer:
    """Tracks content hashes across requests to figure out what's stable
    (same every call) vs what changes (user query, conversation tail).

    The idea: if a content block appears in 2+ of your last N requests,
    it's almost certainly stable and worth caching.
    """

    def __init__(self, window_size: int = 100, stability_threshold: int = 2):
        self._seen: LRU[str, int] = LRU(window_size)
        self._threshold = stability_threshold

    def record_and_count(self, content: str) -> int:
        """Record a content block and return how many times we've seen it."""
        h = hash_normalized(content)
        count = (self._seen.get(h) or 0) + 1
        self._seen.set(h, count)
        return count

    def record_object_and_count(self, obj: object) -> int:
        h = hash_object(obj)
        count = (self._seen.get(h) or 0) + 1
        self._seen.set(h, count)
        return count

    def is_stable(self, content: str) -> bool:
        """Check if we've seen this content enough times to consider it stable."""
        h = hash_normalized(content)
        return (self._seen.get(h) or 0) >= self._threshold

    def is_object_stable(self, obj: object) -> bool:
        h = hash_object(obj)
        return (self._seen.get(h) or 0) >= self._threshold

    def clear(self) -> None:
        self._seen.clear()
